capsules/src/mac.rs

```diff
@@ -0,0 +1,365 @@
+//! Implements 802.15.4 MAC device functionality as an abstraction layer over
+//! the raw radio transceiver hardware. The abstraction difference between a MAC
+//! device and a raw radio transceiver is that the MAC devices exposes a
+//! frame-oriented interface to its users, whereas the radio transceiver
+//! transmits raw byte sequences. There is some abstraction breaking here,
+//! though because the following are still implemented at the hardware level:
+//! - CSMA-CA backoff
+//! - FCS generation and verification
+//!
+//! TODO: Encryption/decryption
+//! TODO: Sending beacon frames
+//! TODO: Channel scanning
+
+use core::cell::Cell;
+use kernel::ReturnCode;
+use kernel::common::take_cell::TakeCell;
+use kernel::hil::radio;
+use net::ieee802154::*;
+
+#[derive(Copy, Clone, Eq, PartialEq, Debug)]
+pub struct FrameInfo {
+    // These offsets are relative to buf[radio::PSDU_OFFSET..] so that
+    // the mac frame length is data_offset + data_len
+    mac_payload_offset: usize,
+    data_offset: usize,
+    data_len: usize,
+    security_params: Option<(Security, [u8; 16])>,
+}
+
+impl FrameInfo {
+    pub fn frame_length(&self) -> usize {
+        self.data_offset + self.data_len
+    }
+
+    pub fn remaining_data_capacity(&self, buf: &[u8]) -> usize {
+        buf.len() - radio::PSDU_OFFSET - radio::MFR_SIZE - self.data_offset - self.data_len
+    }
+
+    pub fn append_payload(&mut self, buf: &mut [u8], payload: &[u8]) -> ReturnCode {
```

Comment:
Yep, the old reason for this is now obsolete. Initially, I used a `Frame` too, but then decided to pull out the `FrameInfo` struct to keep lifetimes simple when trying to avoid copies when passing around the buffer in-place for cryptography. Now that I know that a copy will always have to occur for crypto, I've reworked the logic to again use `Frame`.